packages/core/src/config/parse.ts

```diff
@@ -1722,41 +1731,102 @@ const assertValidConstructorArgs = (
+  artifactPaths: ArtifactPaths,
+  integration: Integration,
+  stream: NodeJS.WritableStream,
+  silent: boolean,
+  exitOnFailure: boolean = true
+): Promise<{ [referenceName: string]: string }> => {
   // Resolve all the contract addresses so they can be used handle contract references regardless or ordering.
   const contracts: { [referenceName: string]: string } = {}
   for (const [referenceName, userContractConfig] of Object.entries(
     userConfig.contracts
   )) {
     const { externalProxy, kind } = userContractConfig
 
-    // Set the proxy address to the user-defined value if it exists, otherwise set it to the default proxy
-    // used by ChugSplash.
-    const proxy =
-      externalProxy ||
-      getDefaultProxyAddress(
-        userConfig.options.organizationID,
-        userConfig.options.projectName,
-        referenceName
-      )
+    // if the contract is proxied then we can just return the proxy address
+    if (kind !== 'no-proxy') {
+      // Set the proxy address to the user-defined value if it exists, otherwise set it to the default proxy
+      // used by ChugSplash.
+      const proxy =
+        externalProxy ||
+        getDefaultProxyAddress(
+          userConfig.options.organizationID,
+          userConfig.options.projectName,
+          referenceName
+        )
 
-    contracts[referenceName] =
-      kind !== 'no-proxy'
-        ? proxy
-        : getContractAddress(
-            userConfig.options.organizationID,
-            referenceName,
-            userContractConfig.constructorArgs ?? {},
-            cachedArtifacts[referenceName]
-          )
+      contracts[referenceName] = proxy
+    } else {
+      /**
+       * If the contract is not proxied then we need to resolve the actual contract address.
+       *
+       * If a no-proxy contract includes libraries, then we must resolve any contract references used when defining those libraries.
+       * This is because the library addresses are injected into the bytecode which will impact the create2 address of the contract
+       * that we are attempting to resolve here.
+       *
+       * Due to this dynamic, if a no-proxy contract dependes on a library, then the library must be defined before the contract
+       * in the config or we will not be able to resolve the contract address.
+       */
+
+      // So if the contract uses libraries, then we compile the libraries to resolve the library contract references, and detect any out of order references.
+      const outOfOrderReferences: string[] = []
+      let libraries = userContractConfig.libraries ?? {}
+      if (libraries && Object.entries(libraries).length > 0) {
+        libraries = JSON.parse(
+          Handlebars.compile(JSON.stringify(libraries ?? {}))({
+            ...contracts,
+          })
+        )
+
+        for (const [libraryName, libraryAddress] of Object.entries(libraries)) {
+          if (libraryAddress === '' && libraries) {
```

Comment:
I don't think `&& libraries` is necessary because we know it's an object that contains at least one element